### ISSUE TITLE
Bad 3D rendering for very small volumes

### ISSUE THREAD

**Issue by jni** [napari/napari#3550]
## 🐛 Bug

3D rendering of few-pixel volumes is atrocious (... under some circumstances?).

```python
import numpy as np
import napari

image = np.array(
      [[[1, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0]],

       [[0, 0, 0, 0, 0],
        [0, 1, 0, 0, 1],
        [0, 0, 0, 0, 1],
        [0, 0, 1, 0, 1],
        [1, 1, 0, 1, 0]],

       [[0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 1, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0]],

       [[0, 0, 0, 0, 0],
        [0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 1, 0, 0, 0]],

       [[0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 1, 0, 0, 1],
        [1, 0, 1, 0, 1],
        [0, 0, 0, 0, 1]]])

viewer = napari.Viewer(ndisplay=3)
image_layer = viewer.add_image(image, interpolation3d='nearest')
viewer.camera.angles = (
        10.278731497835434, 32.352926664548114, 136.03311454553256
        )

napari.run()
```

<img width="1754" alt="Screen Shot 2021-10-28 at 12 16 29 pm" src="https://user-images.githubusercontent.com/492549/139169621-c36888ef-5576-4620-aa2a-b6f8f470cf33.png">

This is unchanged by #3494 + vispy 0.9.3.

napari info:

```
napari: 0.4.12.dev18+gc53f34e85
Platform: macOS-11.4-x86_64-i386-64bit
System: MacOS 11.4
Python: 3.9.7 | packaged by conda-forge | (default, Sep  2 2021, 17:58:46)  [Clang 11.1.0 ]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.20.3
SciPy: 1.7.1
Dask: 2021.09.0
VisPy: 0.9.3

OpenGL:
  - GL version:  2.1 ATI-4.5.14
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 2304x1296, scale 2.0
  - screen 2: resolution 1792x1120, scale 2.0

Plugins:
  - animation: 0.0.2
  - console: 0.0.4
  - napari-midi: 0.1.dev5+g16c91f6.d20210914
  - napari-ome-zarr: 0.3.0
  - nd2-reader: 0.0.6
  - platelet-unet-watershed: 0.0.3
  - scikit-image: 0.4.12.dev18+gc53f34e85
  - svg: 0.1.5
  - tracks: 0.0.2
  - zarpaint: 0.2.0
```

**Comment by brisvag** [comment]
I noticed the same. ~It only happens~ *It's only obvious* using nearest interpolation. It's probably something to do with the amount of steps in the raycasting. You'd probably want to do more of them when the volume is very small?

**Comment by brisvag** [comment]
Yup, I was right! vispy exposes an attribute called `relative_step_size` which determines the step size in texture pixels. By default, this is 0.8. This is what happens if you change it from 0.8 to 0.1:

https://user-images.githubusercontent.com/23482191/139228862-745d8d21-9939-4127-84f6-c62d47d9e1a8.mp4

Obviously this has a (linear?) performance cost. We should probably expose it somehow in napari as a "quality" parameter.

**Comment by jni** [comment]
Awesome! Thanks for the investigation @brisvag! I might go further — could we set `relative_step_size` automatically based on (for example) canvas size relative to data size?

**Comment by brisvag** [comment]
Seems doable, yeah! Not sure exactly where that would live... I think we can access the parent canvas from the visual!

**Comment by alisterburt** [comment]
Super nice! Worth noting that if we do set `relative_step_size` dynamically it may end up looking weird with perspective and we should make sure to test that

**Comment by tlambert03** [comment]
> vispy exposes an attribute called relative_step_size which determines the step size in texture pixels.

woah!  I've wondered about this sawtooth phenomenon for a while!

**Comment by brisvag** [comment]
> Super nice! Worth noting that if we do set `relative_step_size` dynamically it may end up looking weird with perspective and we should make sure to test that

Why would it look odd with perspective specifically?

**Comment by alisterburt** [comment]
@brisvag the relationship just gets a bit more complex, things closer become larger so might need a smaller step size than we would expect based on just data size and canvas size? Doesn't really need to be thought about now, just checked if we do go that route 🙂

**Comment by psobolewskiPhD** [comment]
If someone runs into this, there is a fix widget here:
https://github.com/vispy/vispy/pull/2587#issuecomment-2132678092
@jni I know the adaptive quality https://github.com/napari/napari/pull/6930 has stalled, but maybe just a quality setting could be exposed in the API?

**Comment by jni** [comment]
yes and yes. 😊
